src/decode-nsh.c

```diff
@@ -0,0 +1,337 @@
+/* Copyright (C) 2015 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \ingroup decode
+ *
+ * @{
+ */
+
+
+/**
+ * \file
+ *
+ * \author Carl Smith <carl.smith@alliedtelesis.co.nz>
+ *
+ * Decodes Network Service Header (NSH)
+ */
+
+#include "suricata-common.h"
+#include "suricata.h"
+#include "decode.h"
+#include "decode-events.h"
+#include "decode-nsh.h"
+
+#include "util-unittest.h"
+#include "util-debug.h"
+
+/**
+ * \brief Function to decode NSH packets
+ */
+
+int DecodeNSH(ThreadVars *tv, DecodeThreadVars *dtv, Packet *p, const uint8_t *pkt, uint32_t len)
+{
+    StatsIncr(tv, dtv->counter_nsh);
+
+    /* Check mimimum header size */
+    if (len < sizeof(NshHdr)) {
+        ENGINE_SET_INVALID_EVENT(p, NSH_HEADER_TOO_SMALL);
+        return TM_ECODE_FAILED;
+    }
+
+    /* Sanity check the header version */
+    const NshHdr *hdr = (const NshHdr *)pkt;
+    uint16_t version = ntohs(hdr->ver_flags_len) >> 14;
+    if (version != 0) {
+        ENGINE_SET_EVENT(p, NSH_UNSUPPORTED_VERSION);
+        return TM_ECODE_OK;
+    }
+
+    /* Should always be some data after the header */
+    uint16_t length = (ntohs(hdr->ver_flags_len) & 0x003f) * 4;
+    if (length >= len) {
+        ENGINE_SET_INVALID_EVENT(p, NSH_BAD_HEADER_LENGTH);
+        return TM_ECODE_FAILED;
+    }
+
+    /* Check for valid MD types */
+    uint8_t md_type = hdr->md_type;
+    if (md_type == 0 || md_type == 0xF) {
+        /* We should silently ignore these packets */
+        ENGINE_SET_EVENT(p, NSH_RESERVED_TYPE);
+        return TM_ECODE_OK;
+    } else if (md_type == 1)  {
+        /* Fixed header length format */
+        if (length != 24) {
+            ENGINE_SET_INVALID_EVENT(p, NSH_BAD_HEADER_LENGTH);
+            return TM_ECODE_FAILED; 
+        }
+    } else if (md_type != 2) {
+        /* Not variable header length either */
+        ENGINE_SET_EVENT(p, NSH_UNSUPPORTED_TYPE);
+        return TM_ECODE_OK;
+    }
+
+    /* Now we can safely read the rest of the header */
+    uint8_t next_protocol = hdr->next_protocol;
+#ifdef DEBUG
```

Comment:
Suggest instead
```
if (SCLogDebugEnabled()) {
    uint32_t spi_si = ....
    SCLogDebug("...");
}
```

Reply:
I think I still need to put the ifdef around this. Otherwise you get "unused variable" warnings in a non debug build. That appears to be what others have done for this case. So I will do both.